pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,424 @@
+		Bucket: awssdk.String(bucketName),
+		Tagging: &s3.Tagging{
+			TagSet: []*s3.Tag{
+				{
+					Key:   awssdk.String("kubernetes.io/cluster/" + r.infrastructureName),
+					Value: awssdk.String("owned"),
+				},
+			},
+		},
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			return aerr
+		}
+		return err
+	}
+	r.logger.WithField("bucket", bucketName).Debug("bucket tagged")
+
+	// Render and create the OIDC discovery JSON s3 object in the well-known location
+	issuerURL := r.getIssuerURL()
+	discoveryJSON := fmt.Sprintf(discoveryTemplate, issuerURL, issuerURL, keysURI)
+	_, err = awsClient.PutObject(&s3.PutObjectInput{
+		ACL:    awssdk.String("public-read"),
+		Body:   awssdk.ReadSeekCloser(strings.NewReader(discoveryJSON)),
+		Bucket: awssdk.String(bucketName),
+		Key:    awssdk.String(discoveryURI),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			return aerr
+		}
+		return err
+	}
+	r.logger.WithField("bucket", bucketName).Debug("discovery document updated")
+
+	// Extract the token signer and create keys JSON s3 object for jwks_uri
+	tokenSignerConfigMap, err := r.kubeclientset.CoreV1().ConfigMaps(tokenSignerConfigMapNamespace).Get(context.TODO(), tokenSignerConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	pemKeys := [][]byte{}
+	for _, pemKey := range tokenSignerConfigMap.Data {
+		pemKeys = append(pemKeys, []byte(pemKey))
+	}
+	if len(pemKeys) < 1 {
+		return fmt.Errorf("no signing keys found in config map %s/%s", tokenSignerConfigMapNamespace, tokenSignerConfigMapName)
+	}
+
+	keysJSON, err := encodeKeys(pemKeys)
+	_, err = awsClient.PutObject(&s3.PutObjectInput{
+		ACL:    awssdk.String("public-read"),
+		Body:   awssdk.ReadSeekCloser(bytes.NewReader(keysJSON)),
+		Bucket: awssdk.String(bucketName),
+		Key:    awssdk.String(keysURI),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			return aerr
+		}
+		return err
+	}
+	r.logger.WithField("bucket", bucketName).Debug("keys document updated")
+
+	return nil
+}
+
+// Below this line based on https://github.com/openshift/aws-pod-identity-webhook/blob/master/hack/self-hosted/main.go
+
+// copied from kubernetes/kubernetes#78502
+func keyIDFromPublicKey(publicKey interface{}) (string, error) {
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize public key to DER format: %v", err)
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	return keyID, nil
+}
+
+type KeyResponse struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+func encodeKeys(pemKeys [][]byte) ([]byte, error) {
+	var keys []jose.JSONWebKey
+	var response []byte
+	for _, key := range pemKeys {
+		block, _ := pem.Decode(key)
+		if block == nil {
+			return response, errors.New("Error decoding PEM key")
```

Comment:
do we want to return return response, or should we explicity return nil with err

Reply:
`response` is not a pointer atm